lib/tapioca/compilers/dsl/config.rb

```diff
@@ -0,0 +1,104 @@
+# typed: strict
+# frozen_string_literal: true
+
+begin
+  require "config"
+rescue LoadError
+  return
+end
+
+module Tapioca
+  module Compilers
+    module Dsl
+      # `Tapioca::Compilers::Dsl::Config` generates RBI files for classes generated by the
+      # [`config`](https://github.com/rubyconfig/config) gem.
+      #
+      # The gem creates a `Config::Options` instance based on the settings files and/or
+      # env variables. It then assigns this instance to a constant with a configurable name,
+      # by default `Settings`. Application code uses methods on this constant to read off
+      # config values.
+      #
+      # For a setting file like the following:
+      # ```yaml
+      # ---
+      # github:
+      #   token: 12345
+      #   client_id: 54321
+      #   client_secret: super_secret
+      # ```
+      # and a `Config` setup like:
+      # ```ruby
+      # Config.setup do |config|
+      #   config.const_name = "AppSettings"
+      # end
+      # ```
+      # this generator will produce the following RBI file:
+      # ```rbi
+      # AppSettings = T.let(T.unsafe(nil), AppSettingsConfigOptions)
+      #
+      # class AppSettingsConfigOptions < ::Config::Options
+      #   sig { returns(T.untyped) }
+      #   def github; end
+      #
+      #   sig { params(value: T.untyped).returns(T.untyped) }
+      #   def github=(value); end
+      # end
+      # ```
+      class Config < Base
+        extend T::Sig
+
+        CONFIG_OPTIONS_SUFFIX = "ConfigOptions"
+
+        sig { override.params(root: Parlour::RbiGenerator::Namespace, constant: Module).void }
+        def decorate(root, constant)
+          option_class_name = constant.name
+          return unless option_class_name
+
+          constant_name = option_class_name
+            .gsub(/#{CONFIG_OPTIONS_SUFFIX}$/, "")
+          method_names = constant.instance_methods(false)
+
+          return if method_names.empty?
+
+          root.create_constant(constant_name, value: "T.let(T.unsafe(nil), #{constant.name})")
+
+          root.create_class(option_class_name, superclass: "::Config::Options") do |mod|
+            mod.create_extend("T::Generic")
+            mod.create_constant("Elem", value: "type_member(fixed: T.untyped)")
```

Comment:
Is `Elem` used anywhere?

Reply:
Since `Config::Options` is enumerable, we need to redefine the generic type member in the subclass. That's why I have to declare `Elem` here, so it is only used if someone wants to use `Settings` as an enumerable, each entry will be `T.untyped`. Otherwise, it is not directly used.